File: integration/configuration.py
"""
Variables that influence testing behavior are defined here.
"""

import os
import random

from helpers import crypto
from helpers.utils import merge

try:
    from local_configuration import CONF_OVERRIDE
except ImportError:
    CONF_OVERRIDE = {}


def get():
    """
    Get configuration (non-idempotent function)
    """
    # This "factor" is later used in the network name and port forwarding.
    port_factor = random.randint(0, 4096)
    # Docker network name. Also used as a project and domain name.
    net_name = 'test_net_{num}'.format(num=port_factor)

    dynamic_config = generate_dynamic_config(net_name)

    config = {
        # Common conf options.
        # See below for dynamic stuff (keys, certs, etc)
        'dynamic': dynamic_config,
        # Controls whether to perform cleanup after tests execution or not.
        'cleanup': True,
        # Code checkout
        # Where does all the fun happens.
        # Assumption is that it can be safely rm-rf`ed later.
        'staging_dir': 'staging',
        # Default repository to pull code from.
        'git_repo_base': os.environ.get('DBAAS_INFRA_REPO_BASE', 'ssh://{user}@gerrit.cmail.yandex.net:9501'),
        # Controls whether overwrite existing locally checked out
        # code or not (default)
        'git_clone_overwrite': False,
        # Docker-related
        'docker_ip4_subnet': '10.%s.0/24',
        'docker_ip6_subnet': 'fd00:dead:beef:%s::/96',
        # See above.
        'port_factor': port_factor,
        # These docker images are brewed on `docker.prep_images` stage.
        # Options below are passed as-is to
        # <docker_api_instance>.container.create()
        'network_name': net_name,
        # Docker network name. Also doubles as a project and domain name.
        'projects': {
            # Basically this mimics docker-compose 'service'.
            # Matching keys will be used in docker-compose,
            # while others will be ignored in compose file, but may be
            # referenced in any other place.
            'base': {
                # The base needs to be present so templates,
                # if any, will be rendered.
                # It is brewed by docker directly,
                # and not used in compose environment.
                'docker_instances': 0,
            },
            'mongodb': {
                'build': '..',
                # Config can have arbitrary keys.
                # This one is used in template matching of config file options.
                # See Dockerfile itself for examples.
                # 'docker_instances': 3,
                'users': {
                    'admin': {
                        'username': 'admin',
                        'password': 'password',
                        'dbname': 'admin',
                        'roles': ['root'],
                    },
                },
                'expose': {
                    'mongod': 27018,
                    'ssh': 22,
                },
                'docker_instances': 2,
                'external_links': ['%s:%s' % (dynamic_config['s3']['host'], dynamic_config['s3']['fake_host'])],
            },
            'minio': {
                'build': 'images/minio',
                'expose': {
                    'http': 9000,
                },
            },
        },
        # A dict with all projects that are going to interact in this
        # testing environment.
        'base_images': {
            'dbaas-mongodb-backup-base': {
                'tag': 'dbaas-mongodb-backup-base',
                'path': 'images/base',
            },
        },
    }
    return merge(config, CONF_OVERRIDE)


def generate_dynamic_config(net_name):
    """
    Generates dynamic stuff like keys, uuids and other.
    """
    keys = {
        'internal_api': crypto.gen_keypair(),
        'client': crypto.gen_keypair(),
    }
    # https://pynacl.readthedocs.io/en/latest/public/#nacl-public-box
    # CryptoBox is a subclass of Box, but returning a string instead.
    api_to_client_box = crypto.CryptoBox(
        keys['internal_api']['secret_obj'],
        keys['client']['public_obj'],
    )
    s3_credentials = {
        'access_secret_key': crypto.gen_plain_random_string(40),
        'access_key_id': crypto.gen_plain_random_string(20),
    }

    gpg_private_key, gpg_key_id = crypto.gen_gpg('test_cluster')

    config = {
        's3': {
            'host': 'minio01.{domain}'.format(domain=net_name),
            'fake_host': 'minio',
            'fake_host_port': 'minio:9000',
            'bucket': 'dbaas',
            'endpoint': 'http://minio:9000',
            'access_secret_key': s3_credentials['access_secret_key'],
            'access_key_id': s3_credentials['access_key_id'],
            'enc_access_secret_key': api_to_client_box.encrypt_utf(s3_credentials['access_secret_key']),
            'enc_access_key_id': api_to_client_box.encrypt_utf(s3_credentials['access_key_id']),
        },
        'gpg': {
            'private_key': gpg_private_key,
            'key_id': gpg_key_id,
            'user': 'mongodb',
            'homedir': '/home/mongodb/.gnupg',
        },
        'wal-g': {
            'path': 'mongodb-backup/test_uuid/test_mongodb',
        },
    }

    return config

File: tests_func/configuration.go
package main

import (
	"fmt"
	"math/rand"
)

type TempNameType1 struct {
	tag  string
	path string
}

type ConfigurationType struct {
	DynamicConfiguration DynamicConfigurationType
	Cleanup              bool
	StagingDir           string
	ImagesDir            string
	GitCloneOverwrite    bool
	DockerBridgeName     string
	DockerBridgeId       int
	DockerIP4Subnet      string
	DockerIP6Subnet      string
	PortFactor           string
	NetworkName          string
	Projects             map[string]DockerComposeConfiguration
	BaseImages           map[string]TempNameType1
}

type UserConfiguration struct {
	Username string
	Password string
	Dbname   string
	Roles    []string
}

type DockerComposeConfiguration struct {
	Build           string
	Users           map[string]UserConfiguration
	Expose          map[string]int
	DockerInstances int
	ExternalLinks   []string
}

func getConfiguration(testContext *TestContextType) ConfigurationType {
	portFactor := GetVarFromEnvList(testContext.Env, "TEST_ID")
	netName := fmt.Sprintf("test_net_%s", portFactor)
	dynamicConfig := getDynamicConfiguration(testContext)
	configuration := ConfigurationType{
		DynamicConfiguration: dynamicConfig,
		Cleanup:              true,
		StagingDir:           "staging",
		ImagesDir:            "images",
		GitCloneOverwrite:    false,
		DockerBridgeName:     "dbaas",
		DockerBridgeId:       rand.Intn(65535),
		DockerIP4Subnet:      "10.%s.0/24",
		DockerIP6Subnet:      "fd00:dead:beef:%s::/96",
		PortFactor:           portFactor,
		NetworkName:          netName,
		Projects: map[string]DockerComposeConfiguration{
			"base": {DockerInstances: 0},
			"mongodb": {
				Build: "..",
				Users: map[string]UserConfiguration{
					"admin": {
						Username: "admin",
						Password: "password",
						Dbname:   "admin",
						Roles:    []string{"root"},
					},
				},
				Expose: map[string]int{
					"mongod": 27018,
					"ssh":    22,
				},
				DockerInstances: 2,
				ExternalLinks:   []string{dynamicConfig.s3.host, dynamicConfig.s3.fakeHost},
			},
			"minio": {
				Build:  "staging/images/minio",
				Expose: map[string]int{"http": 9000},
			},
		},
		BaseImages: map[string]TempNameType1{
			"mongodb-backup-base": {
				tag:  "mongodb-backup-base",
				path: "staging/images/base",
			},
		},
	}
	return configuration
}

type S3Configuration struct {
	host               string
	fakeHost           string
	fakeHostPort       string
	bucket             string
	endpoint           string
	accessSecretKey    string
	accessKeyId        string
	encAccessSecretKey string
	encAccessKeyId     string
}

type GPGConfiguration struct {
	privateKey string
	keyId      string
	user       string
	homedir    string
}

type WalgConfiguration struct {
	path string
}

type DynamicConfigurationType struct {
	s3   S3Configuration
	gpg  GPGConfiguration
	walg WalgConfiguration
}

func getDynamicConfiguration(testContext *TestContextType) DynamicConfigurationType {
	return DynamicConfigurationType{
		s3: S3Configuration{
			host:               fmt.Sprintf("minio01.%s", GetVarFromEnvList(testContext.Env, "TEST_ID")),
			fakeHost:           "minio",
			fakeHostPort:       "minio:9000",
			bucket:             "dbaas",
			endpoint:           "http://minio:9000",
			accessSecretKey:    GetVarFromEnvList(testContext.Env, "MINIO_SECRET_KEY"),
			accessKeyId:        GetVarFromEnvList(testContext.Env, "MINIO_ACCESS_KEY"),
			encAccessSecretKey: "TODO",
			encAccessKeyId:     "TODO",
		},
		gpg: GPGConfiguration{
			privateKey: "TODO",
			keyId:      "TODO",
			user:       "mongodb",
			homedir:    "/home/mongodb/.gnupg",
		},
		walg: WalgConfiguration{path: GetVarFromEnvList(testContext.Env, "WALG_S3_PREFIX")},
	}
}
